[Recorder] Docs update (#33571)
Azure/azure-sdk-for-js: sdk/test-utils/recorder/ASSET_SYNC_WORKFLOW.md

- [Test-Proxy Commands](#test-proxy-commands)
- [Working Offline](#working-offline)
- [Troubleshooting](#troubleshooting)
- [Frequently Asked Questions](#frequently-asked-questions)
- [Further Reading](#further-reading)

## Background

Recordings take up a large amount of space in our repository and generate a lot of churn. The asset sync project, owned by our Engineering System team, is a solution to this problem. Leveraging and extending the existing test proxy, the asset sync extension adds support for moving the recordings outside of the main azure-sdk-for-js repo.

## Prerequisites

To be able to leverage the asset-sync workflow:

- **Repository access**:
  - You need **write-access** to the `azure-sdk-assets` repo to push recordings
  - [Request permissions here](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/785/Externalizing-Recordings-(Asset-Sync)?anchor=permissions-to-%60azure/azure-sdk-assets%60)
- **Tools**:
  - Install [Powershell](https://github.com/PowerShell/PowerShell)
  - Verify the `pwsh` command works in your terminal
- **Dependencies**:
  - Add `@azure/dev-tool` to the `devDependencies` in your `package.json`

>   - Ensure your package uses `@azure-tools/test-recorder@^4.0.0` or later

Reviewer: ```suggestion
  - Ensure your package has a dev dependency on `@azure-tools/test-recorder@^4.0.0` or later
```

## New Package - No Recorded Tests

This section is for packages that are new to the JS repo and are trying to onboard tests with `@azure-tools/test-recorder` and the asset-sync workflow.

From the root of the repo, navigate to your package:

```
cd sdk/<service-folder>/<package-name>
```

Generate an `sdk/<service-folder>/<package-name>/assets.json` file by running the following command:

```
npx dev-tool test-proxy init
```

> This command will generate an `assets.json` file with an empty tag.

Reviewer 2: Should we make it clear to git commit that file?

Author: I’ve added that where we talk about tag getting populated.

Author: "
You should stage and commit the `assets.json` update as part of your PR. If you don't run the `push` command before creating a PR, the CI (and anyone else who tries to run your recorded tests) will use the old recordings, which will cause failures.
"

1. Push your local recordings to a tag in the `Azure/azure-sdk-assets` repo, and
2. Update the `assets.json` in your package root to reference the newly created tag.

Example `assets.json` from "keyvault-certificates" SDK.

```json
{
  "AssetsRepo": "Azure/azure-sdk-assets",
  "AssetsRepoPrefixPath": "js",
  "TagPrefix": "js/keyvault/keyvault-certificates",
  "Tag": "js/keyvault/keyvault-certificates_43821e21b3"
}
```

And the recordings are located at https://github.com/Azure/azure-sdk-assets/tree/js/keyvault/keyvault-certificates_43821e21b3

You should stage and commit the `assets.json` update as part of your PR. If you don't run the `push` command before creating a PR, the CI (and anyone else who tries to run your recorded tests) will use the old recordings, which will cause failures.

After onboarding your new package or after migrating your package to the asset-sync workflow, the following diagram describes the new workflow (new steps highlighted):

```mermaid
graph TD
    subgraph p3 [Asset Sync]
        subgraph p4 [Push to Assets Repo]
            push[npx dev-tool test-proxy push]
        end
        p4-->assets[Commit assets.json]
    end

    assets --> pr[Create PR]

>     subgraph p2 [Verify in Playback]

Author: ```suggestion
    subgraph p2 [Verify in Playback <br> (check it works without .env file/vars)]
```

>     subgraph p2 [Verify in Playback <br> (check it works without .env file/vars)]

Author: ```suggestion
    subgraph p2 [Verify in Playback without .env file/vars]
```

>         playback[TEST_MODE=playback<br>rushx test]

Reviewer 2: I think it would be more accurate to verify it works in the absence of .env file. From experience, it is easy for tests to accidentally depend on env vars in playback mode.

Author: As long as you set TEST_MODE to playback, we don’t care about the env variables in the recorder and instead pick the fake secrets. If that’s not happening, let’s sync up offline to see how to make the experience better.

Author: added anyway

    p1 --> p2
    p2 -- Tests don't work re-record --> p1

>     p2 -- Tests pass in playback; and are properly sanitized --> p4

Author: New line was not rendering. Not sure if it was that way previously.

\n was showing up in text.

Anyways, I’ve updated to have line breaks <br> instead, which is working.

**Refer to [testing-commands](https://github.com/Azure/azure-sdk-for-js/wiki/Golden-Testing-Commands) guide if you need help on the commands to run during testing.**

## Working Offline

Offline work is supported out-of-the-box. Of course, however, you won't be able to push or pull from the assets repo while offline. You can fetch recordings from the assets repo by running `npx dev-tool test-proxy restore`. This will download the recordings (and the test proxy executable, if you haven't got that already), making them ready for you to run tests with.

## Troubleshooting

Common issues and their solutions:

### Error: Permission denied to azure-sdk-assets repository
- Verify you have write access to the `Azure/azure-sdk-assets` repository
- Check that your Git credentials are properly configured

### Error: Powershell command not found
- Ensure Powershell is installed and the `pwsh` command is in your `PATH`.
- Try reinstalling Powershell if necessary

### Tests fail in playback mode after migration

> - Check that your `assets.json` was properly committed and contains the correct tag

Reviewer 3: ```suggestion
- Check that your `assets.json` is properly committed and contains the correct tag
```

- Ensure Powershell is installed and the `pwsh` command is in your `PATH`
- Try reinstalling Powershell if necessary

### Tests fail in playback mode after migration
- Check that your `assets.json` is properly committed and contains the correct tag
- Try running `npx dev-tool test-proxy restore` to ensure you have the latest recordings
- Examine the error logs for specific sanitization or playback issues

### Changes not appearing in recordings
- Verify you're running tests in record mode with `TEST_MODE=record`
- Run `npx dev-tool test-proxy push` after recording
- Check if the changes appear in the `.assets` directory

### Tests work in CI but fail locally
If tests run in playback mode successfully in the CI pipelines but fail on your local machine, it is likely that your local environment or recordings might be corrupted
- **Reset your recordings**: Delete the `.assets` folder in the root of the repository to remove all pulled recordings and give yourself a clean slate
- Then run `npx dev-tool test-proxy restore` to pull fresh copies of the recordings

This is particularly helpful when you've been switching branches or if there have been significant recording changes

> ### Outdated test-proxy version

Reviewer 4: Has this situation ever actually happened? The test proxy download script looks at the version in eng/common first and should automatically download that version if it doesn't match the currently downloaded version. So if the current version is outdated, it will get removed and the new version downloaded automatically.

The only time I can think of off the top of my head where removing the downloaded test proxy would help would be if the executable got corrupted somehow.

Author: That makes sense. I wanted to give this option as I remember encountering such a situation.

Now that I think more on the situation, maybe it was such that a new release had a regression and I wanted to use a different version.

Removing this section for now. Will add similar ones if we encounter in the future.

> ### Outdated test-proxy version
> If you're experiencing issues that might be related to an older version of test-proxy, check your test logs for the installation location
> - Look for log lines like: `[test-proxy] Test proxy executable already exists at /home/codespace/.cache/azsdk-dev-tool/1.0.0-dev.20250221.1/Azure.Sdk.Tools.TestProxy, not downloading it.`
> - Delete the folder containing the test-proxy executable (in the example above, delete the `/home/codespace/.cache/azsdk-dev-tool/1.0.0-dev.20250221.1/` directory)
> - This will force a fresh download of the test-proxy when you run your tests again
> 
> This is especially helpful if there are recent bug fixes in the test-proxy that you need
> 

Author: ```suggestion
```

remove
